### PR TITLE
Improve output when function fails in pipe

### PR DESCRIPTION
**Pull request by Shell32-Natsu**
When there is any error in function:

1. The function failure message `[FAIL] "gcr.io/kpt-fn/xxx"` will be output to stderr.
2. Structured results will be output to stderr.
3. Content in stderr and exit code will be output to stderr.

The output of piped eval command
```
kpt fn source \
| kpt fn eval - --image gcr.io/kpt-fn/set-namespace:v0.1 \
| kpt fn eval - --image gcr.io/kpt-fn/dne -- foo=bar \
| kpt fn sink .
```
changed from 
```
error: Stderr:
    "failed to configure function: input namespace cannot be empty"
  Exit Code: 1
error: Function image "gcr.io/kpt-fn/dne" doesn't exist.
```
to
```
[FAIL] "gcr.io/kpt-fn/set-namespace:v0.1"
  Results:
    [ERROR] failed to configure function: input namespace cannot be empty
  Stderr:
    "failed to configure function: input namespace cannot be empty"
  Exit code: 1


[FAIL] "gcr.io/kpt-fn/dne"
error: Function image "gcr.io/kpt-fn/dne" doesn't exist.
```

#2010

### CODE REVIEW

File: e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+exitCode: 1
+stdErr: |
+  error:   Stderr:
```
Comment (frankfarzan) [inline e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:17]: What's the equivalent output when not piping?
Why/how this differs from that output?

Comment (Shell32-Natsu) [reply e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:17]: If there is only the first function in pipe, the output will be:
```
Stderr:
  "failed to configure function: input namespace cannot be empty"
Exit Code: 1
```
I guess we should output structured result to stderr when there is an error in the function in this case?

Comment (droot) [reply e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:17]: Had an offline chat with @Shell32-Natsu about it. I think we can bring in consistency in the error output in `piped` and `non-piped` scenario. So if we have multiple `fn eval` chained, the error output need to have enough context such as "which function" and all the details such as structured-result output, stderr output dumped on `stderr` (something we already do in case of `non-piped`). He is going to share the proposed output. It will require minor tweak in our printing logic (`stderr` vs `stdout`).

Comment (Shell32-Natsu) [reply e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:17]: Expected output for kubeval:
```
[FAIL] "gcr.io/kpt-fn/kubeval:unstable"
  Results:
    [ERROR] selector is required in object "apps/v1/Deployment/" in file "resources.yaml" in field "selector"
    [ERROR] template is required in object "apps/v1/Deployment/" in file "resources.yaml" in field "template"
    [ERROR] Invalid type. Expected: [integer,null], given: string in object "apps/v1/Deployment/" in file "resources.yaml" in field "spec.replicas"
    [ERROR] Validating arbitrary CRDs is not supported yet. You can skip them by setting ignore_missing_schemas or skip_kinds in the function config
: ERR  - stdin: Failed initializing schema file:///tmp/master-standalone/custom-custom-v1.json: open /tmp/master-standalone/custom-custom-v1.json: n
o such file or directory  in object "custom.io/v1/Custom/" in file "resources.yaml"
  Exit code: 1
```

For set-namespace:
```
[FAIL] "gcr.io/kpt-fn/set-namespace:v0.1"
  Results:
    [ERROR] failed to configure function: input namespace cannot be empty
  Stderr:
    "failed to configure function: input namespace cannot be empty"
  Exit code: 1
```

WDYT? @frankfarzan @droot

File: e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  [FAIL] "gcr.io/kpt-fn/set-namespace:v0.1"
+    Results:
+      [ERROR] failed to configure function: input namespace cannot be empty
+    Stderr:
+      "failed to configure function: input namespace cannot be empty"
+    Exit code: 1
+
+   
+  [FAIL] "gcr.io/kpt-fn/dne"
+  error: Function image "gcr.io/kpt-fn/dne" doesn't exist.
```
Comment (frankfarzan) [inline e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:26]: This seems inconsistent compared to above. I understand it's the orchestrator throwing the error, but we should think of a way to present this better.

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml [hunk 3]
```diff
@@ -18,4 +18,3 @@ image: gcr.io/kpt-fn/set-namespace:v0.1
 stdErr: "failed to configure function: input namespace cannot be empty"
 stdOut: |
   [RUNNING] "gcr.io/kpt-fn/set-namespace:v0.1"
-  [FAIL] "gcr.io/kpt-fn/set-namespace:v0.1"
```
Comment (frankfarzan) [inline e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:21]: Why this change?

Also, I've commented on this in couple of other PRs. What's the rationale for what's printed to stderr vs stdout? Why not just use stderr like when piping? Also if you print to different streams, then they may be presented out of order.

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml (hunk 3 above)
Comment (Shell32-Natsu) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:21]: > Why this change?

I had a discussion with @droot and we agreed that the line `[FAIL] ...` is an error message and should be printed to stderr to provide enough information about which function fails in a command pipe.

> Also, I've commented on this in couple of other PRs. What's the rationale for what's printed to stderr vs stdout? Why not just use stderr like when piping? Also if you print to different streams, then they may be presented out of order.

The detail design for what should be printed to stderr and what should be printed to stdout should be done in #1915. We can have tests to cover the behavior before #1915 is done.

File: e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml (hunk 2 above)
Comment (Shell32-Natsu) [reply e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:26]: I think we should address all these issue in #1915

Comment (Shell32-Natsu) [reply e2e/testdata/fn-eval/error-in-pipe/.expected/config.yaml:26]: There is another issue https://github.com/GoogleContainerTools/kpt/issues/2084 which is relative to this error message.

File: e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml (hunk 3 above)
Comment (frankfarzan) [reply e2e/testdata/fn-eval/missing-fn-config/.expected/config.yaml:21]: ok. I think using two channels is not the way to go for the reasons I stated. So consider it designed. Won't block this PR.

**Review by frankfarzan — APPROVED** [review]
